src/client/streamableHttp.ts

```diff
@@ -0,0 +1,370 @@
+        headers.set('last-event-id', this._lastEventId);
+      }
+
+      const response = await fetch(this._url, {
+        method: 'GET',
+        headers,
+        signal: this._abortController?.signal,
+      });
+
+      if (response.status === 405 || response.status === 404) {
+        // Server doesn't support GET for SSE, which is allowed by the spec
+        // We'll rely on SSE responses to POST requests for communication
+        return;
+      }
+
+      if (!response.ok) {
+        if (response.status === 401 && this._authProvider) {
+          // Need to authenticate
+          return await this._authThenStart();
+        }
+
+        const error = new Error(`Failed to open SSE stream: ${response.status} ${response.statusText}`);
+        this.onerror?.(error);
+        throw error;
+      }
+
+      // Successful connection, handle the SSE stream as a standalone listener
+      const streamId = `initial-${Date.now()}`;
+      this._handleSseStream(response.body, streamId);
+    } catch (error) {
+      this.onerror?.(error as Error);
+      throw error;
+    }
+  }
+
+  async start() {
+    if (this._activeStreams.size > 0) {
+      throw new Error(
+        "StreamableHTTPClientTransport already started! If using Client class, note that connect() calls start() automatically.",
+      );
+    }
+
+    this._abortController = new AbortController();
+    return await this._startOrAuth();
+  }
+
+  /**
+   * Call this method after the user has finished authorizing via their user agent and is redirected back to the MCP client application. This will exchange the authorization code for an access token, enabling the next connection attempt to successfully auth.
+   */
+  async finishAuth(authorizationCode: string): Promise<void> {
+    if (!this._authProvider) {
+      throw new UnauthorizedError("No auth provider");
+    }
+
+    const result = await auth(this._authProvider, { serverUrl: this._url, authorizationCode });
+    if (result !== "AUTHORIZED") {
+      throw new UnauthorizedError("Failed to authorize");
+    }
+  }
+
+  async close(): Promise<void> {
+    // Close all active streams
+    for (const reader of this._activeStreams.values()) {
+      try {
+        reader.cancel();
+      } catch (error) {
+        this.onerror?.(error as Error);
+      }
+    }
+    this._activeStreams.clear();
+
+    // Abort any pending requests
+    this._abortController?.abort();
+
+    // If we have a session ID, send a DELETE request to explicitly terminate the session
+    if (this._sessionId) {
+      try {
+        const commonHeaders = await this._commonHeaders();
+        const response = await fetch(this._url, {
+          method: "DELETE",
+          headers: commonHeaders,
+          signal: this._abortController?.signal,
+        });
+
+        if (!response.ok) {
+          // Server might respond with 405 if it doesn't support explicit session termination
+          // We don't throw an error in that case
```

Comment:
You might also see a 404 if the session ID has already been deleted on the server side (because the server could clear it at any time).